### ISSUE TITLE
Implement async return analysis

### ISSUE THREAD

**Issue by hach-que** [RedpointGames/llvm-project#10]
I won't get a chance to implement this now before Games Week, so I wanted to document the design of this for later.

## What we want the static analyser to catch

Consider the following code:

```cpp
void FClass::SomeFunction(int A, int B, FSimpleDelegate CompletionCallback)
{
  // ... do work here
  // forget to do CompletionCallback.ExecuteIfBound().
}
```

When this happens, code that is calling `SomeFunction` will appear to stall indefinitely. Because there's no first-class async/await in the version of C++ that Unreal Engine targets, it is extremely difficult to track down exactly which function is missing a call to it's completion callback because the function at fault could execute many frames after the source of the original operation.

## Adding a new C++ attribute

The idea is to add a new C++ attribute that can be used to annotate parameters and local variable declarations; `[[redpoint::async_return]]`. This attribute can be associated with parameters and local variable declarations of types `TDelegate`, `TFunction`, lambdas and any type that overloads the `()` operator. 

Static analysis detects this attribute on parameters, and makes the guarantee that for the value annotated with it, either one of two things happens:

- The value is invoked/called via the `()` operator, or by calling a function named `Execute` or `ExecuteIfBound` on it, or
- The value is passed as a parameter to another function call, where that parameter is also marked with `[[redpoint::async_return]]`, or
- The value is captured by a lambda; the lambda itself becomes `[[redpoint::async_return]]` (so the lambda must be called/meet these rules between it's declaration and return of the enclosing function), and the capture value inside the lambda inherits it's `[[redpoint::async_return]]` so that the capture must meet these rules within the lambda body.

It also needs to support various flags:

- `[[redpoint::async_return(if_true)]]`: The value meets the rules above *only* on code paths where the enclosing function returns `true`. It's an error if any of the above conditions are met on a code path that returns `false`. Functions that have a parameter annotated with this mode can't have dynamic `return` statements - they must either be `return true;` or `return false;` for the static analyser to work.
- `[[redpoint::async_return(if_false)]]`: The opposite of `if_true`; the rules must be met if the enclosing function returns `false`.
- `[[redpoint::async_return(no_verify)]]`: Disables static analysis within this function, but allows this parameter to be treated as `[[redpoint::async_return]]` for the purposes of other functions passing values into it. This is intended to be used sparingly and only when you have a function that can't be analysed by the static analyser (e.g. the parameter gets copied to the heap, or crosses some other kind of native boundary, etc.)

## Transitivity for delegate values

For delegates, we need to propagate `[[redpoint::async_return]]` through the `CreateLambda`, etc. functions, so that code such as the following validates correctly:

```cpp
void OtherFunction([[redpoint::async_return]] FSimpleDelegate CompletionCallback)
{
  CompletionCallback.ExecuteIfBound();
}

void SomeFunction([[redpoint::async_return]] FSimpleDelegate CompletionCallback)
{
  // @note: FSimpleDelegate::CreateLambda *also* needs to inherit it's async_return status from the lambda it was passed.
  OtherFunction(FSimpleDelegate::CreateLambda(
    // @note: This lambda value automatically inherits [[redpoint::async_return]] as per the rules above.
    [CompletionCallback]() {
      CompletionCallback.ExecuteIfBound();
    }));
}
```

There's additional complexity around this for delegates using `CreateSP` whereby one of the user parameters of the target function is marked with `[[redpoint::async_return]]` and thus the positional argument of `CreateSP` after the function pointer must be `[[redpoint::async_return]]`

## Special cases

We can't change engine code to use this attribute. As a special case, `AddTicker` on `FTSTicker` is treated as if the delegate parameter was annotated with `[[redpoint::async_return(if_false, no_verify)]]`.

## Virtual functions and overrides

If a class declares a virtual function, and that virtual function has a `[[redpoint::async_return]]` parameter, overrides must also annotate that same parameter with `[[redpoint::async_return]]`:

```cpp
class A {
  virtual void SomeCall([[redpoint::async_return]] FSimpleDelegate CompletionCallback) = 0;
};

class B {
  // @error: CompletionCallback must have the [[redpoint::async_return]] attribute
  virtual void SomeCall(FSimpleDelegate CompletionCallback) override
  {
  }
};
```

## Implementation

Clang provides a `CFG` class, which can be used on a function body to determine all of the possible paths code can flow through it.

For performance reasons, we might need to detect `[[redpoint::async_return]]` inline as the compiler is parsing (rather than afterwards using matchers), building up a list of function declarations to analyse on a background thread later.

**Comment by hach-que** [comment]
I don't plan on pursuing this any more, as it is not necessary with C++20 `co_await`, which all platforms that Unreal Engine 5.5 targets now support at some level.
